src/signature/commitment_signature.rs

```diff
@@ -0,0 +1,292 @@
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+//! Digital Signature module
+//! This module defines generic traits for handling the digital signature operations, agnostic
+//! of the underlying elliptic curve implementation
+
+use crate::{
+    commitment::{HomomorphicCommitment, HomomorphicCommitmentFactory},
+    keys::{PublicKey, SecretKey},
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    cmp::Ordering,
+    hash::{Hash, Hasher},
+    marker::PhantomData,
+    ops::{Add, Mul},
+};
+use tari_utilities::{hex::Hex, ByteArray};
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error, PartialEq, Eq, Deserialize, Serialize)]
+pub enum CommitmentSignatureError {
+    #[error("An invalid challenge was provided")]
+    InvalidChallenge,
+}
+
+#[allow(non_snake_case)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CommitmentSignature<P, K, C> {
+    public_commitment_nonce: HomomorphicCommitment<P>,
+    signature_u: K,
+    signature_v: K,
+    _commitment_factory: PhantomData<C>,
+}
+
+impl<P, K, C> CommitmentSignature<P, K, C>
+where
+    P: PublicKey<K = K>,
+    K: SecretKey,
+    C: HomomorphicCommitmentFactory<P = P> + Default,
+{
+    pub fn new(public_commitment_nonce: HomomorphicCommitment<P>, signature_u: K, signature_v: K) -> Self {
+        CommitmentSignature {
+            public_commitment_nonce,
+            signature_u,
+            signature_v,
+            _commitment_factory: PhantomData,
+        }
+    }
+
+    /// This is the left-hand side of the signature verification equation
+    pub fn calc_signature_verifier(&self) -> HomomorphicCommitment<P> {
```

Comment:
```suggestion
    pub fn calc_signature_verifier(&self, factory: &C) -> HomomorphicCommitment<P> {
```

Reply:
The other approach is to calculate this in the constructor and store the value, providing this as a getter. Depends on how often the verifier is needed on its own

Reply:
I prefer locking it down. Because the `factory` will be related to the `HomomorphicCommitment` which is in turn related to the `public/private keys` Its not something that can change midway. You need to ensure you use the correct one to sign and verify. 

As to the reference, and or using a generic on only the function. Because of the way the code is written, you will need to pass that in every time you sign and or verify. As `calc_signature_verifier` is called by `verify` which is used by `verify_challenge`.  The only functions that don't use the factory are the getters.
So if you need to pass one in every time, why not just lock it down at creation?